sidecar/proxy/monitor/discovery.go

```diff
@@ -113,6 +143,8 @@ func (m *discoveryMonitor) poll() error {
 
 	// Notify the listeners
 	instancesByValue := instanceListAsValues(instances)
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
```

Comment:
Invoking callbacks with a mutex acquired is generally not a good idea (more prone to deadlocks, long running operations, ..). 

Actually, a previous incarnation of the `RegistryMonitor` had to deal with this problem as well, and used copy-on-write method for that: When a listener is added/removed, a _new_ slice is allocated, populated, and assigned to `m.listeners`. When a callback is invoked, the slice _pointer_ is atomically copied from m.listeners, and then the callbacks are invoked while iterating on the local slice pointer, without a lock.